File: backend/src/orchestrator/telemetry.py
"""Telemetry and state snapshot contracts for Part 4 Orchestrator and future frontend."""

from dataclasses import dataclass, field, asdict
from typing import Optional, Tuple, Dict, Any, List
import time


@dataclass
class SystemTelemetry:
    """Comprehensive snapshot of the autonomous FSOC system for logging and frontend telemetry.
    
    Fields required for downstream UI:
    - current_state: Active operational state string (e.g. 'FSOC_ACTIVE', 'OPTICAL_SEARCH')
    - beacon_detected: True if optical beacon spot is presently identified
    - beacon_position: 2D pixel coordinates (x, y) of optical beacon
    - tracking_status: Status string from Part 2 tracker ('LOCKED', 'COASTING', 'SEARCHING', 'LOST')
    - prediction_status: Status of kinematic prediction ('ACTIVE', 'IDLE', 'DEGRADED')
    - optical_recovery_status: Status of optical search and local reacquisition
    - rf_status: Auxiliary RF channel state ('IDLE', 'SCANNING', 'FOUND', 'NO_SIGNAL')
    - terminal_id: Target/identified terminal identifier (e.g. 'TERMINAL_02')
    - authentication_status: Cryptographic authentication state ('UNVERIFIED', 'AUTHENTICATED', 'REJECTED')
    - approximate_direction: Coarse RF bearing angle in degrees [0, 360)
    - alignment_status: Boresight alignment status ('ALIGNED', 'MISALIGNED', 'LOST')
    - fsoc_status: Optical communications link status ('ESTABLISHED', 'SEARCHING', 'DOWN')
    - error_status: Active error description if any
    """
    timestamp: float = field(default_factory=time.time)
    current_state: str = "IDLE"
    beacon_detected: bool = False
    beacon_position: Optional[Tuple[float, float]] = None
    tracking_status: str = "IDLE"
    prediction_status: str = "IDLE"
    optical_recovery_status: str = "IDLE"
    rf_status: str = "IDLE"
    terminal_id: Optional[str] = None
    authentication_status: str = "UNVERIFIED"
    approximate_direction: Optional[float] = None
    alignment_status: str = "LOST"
    fsoc_status: str = "DOWN"
    error_status: Optional[str] = None
    
    # Kinematic & Optical Metrics
    camera_orientation_deg: float = 0.0
    camera_pan: float = 0.0
    camera_tilt: float = 0.0
    pointing_error_px: Optional[float] = None
    pointing_error_deg: Optional[float] = None
    tracking_error: float = 0.0
    rf_rssi: Optional[float] = None
    rf_direction: Optional[float] = None
    tracking_confidence: float = 0.0
    miss_count: int = 0
    recovery_attempt: int = 0
    
    # Real-time Simulation & Terminal Kinematics
    terminal_world_position: Tuple[float, float] = (0.0, 0.0)
    terminal_velocity: Tuple[float, float] = (0.0, 0.0)
    terminal_heading: float = 0.0
    beacon_image_position: Optional[Tuple[float, float]] = None
    predicted_beacon_position: Optional[Tuple[float, float]] = None
    optical_signal_strength: float = 1.0
    optical_link_status: str = "DOWN"
    turbulence_level: str = "LOW"
    vibration_level: float = 0.0
    cloud_occlusion: bool = False
    
    # Terminal & Identification Configuration
    num_terminals: int = 5
    target_terminal: str = "TERMINAL_02"
    target_terminal_name: str = "Terminal-02"
    identification_status: str = "IDENTIFIED"
    
    # Cryptographic Authentication Telemetry (HMAC-SHA256)
    security_algorithm: str = "HMAC-SHA256"
    security_challenge_nonce: str = ""
    security_hmac_sample: str = ""
    security_replay_protected: bool = True
    security_valid: bool = True
    security_stage: str = "AUTHENTICATED"
    
    # High-level demonstration & recovery lifecycle telemetry
    recovery_mechanism: str = "NONE"
    rf_discovery_state: str = "STANDBY"
    rf_discovered_count: int = 0
    rf_total_count: int = 5
    rf_wave_active: bool = False
    optical_search_state: str = "STANDBY"
    optical_tracking_state: str = "STANDBY"
    fsoc_state: str = "STANDBY"
    terminals_overview: List[Dict[str, Any]] = field(default_factory=list)
    
    # Event Log Stream (latest messages)
    recent_events: List[str] = field(default_factory=list)

    def to_dict(self) -> Dict[str, Any]:
        """Convert snapshot to JSON-serializable dictionary for API/WebSocket transmission."""
        d = asdict(self)
        if self.beacon_position is not None:
            d["beacon_position"] = list(self.beacon_position)
        if self.beacon_image_position is not None:
            d["beacon_image_position"] = list(self.beacon_image_position)
        if self.predicted_beacon_position is not None:
            d["predicted_beacon_position"] = list(self.predicted_beacon_position)
        d["terminal_world_position"] = list(self.terminal_world_position)
        d["terminal_velocity"] = list(self.terminal_velocity)
        return d

File: src/types/telemetry.ts
/**
 * Wire contract for the Python mission-control backend (`src/web/server.py`).
 *
 * These types mirror the JSON emitted by `/api/telemetry` and `/api/stream`
 * exactly. Fields the simulator leaves unset arrive as `null` rather than being
 * omitted, so they are typed `T | null` instead of optional.
 */

/** Cartesian pair in arena world pixels (1280x720). */
export type Vec2 = readonly [number, number];

/** Operational states of the Part 4 recovery state machine. */
export type SystemState =
  | 'IDLE'
  | 'RF_DISCOVERY'
  | 'RF_AUTHENTICATION'
  | 'RF_DIRECTION_RECOVERY'
  | 'OPTICAL_SEARCH'
  | 'OPTICAL_TRACKING'
  | 'OPTICAL_REACQUISITION'
  | 'PREDICTIVE_RECOVERY'
  | 'LOCAL_REACQUISITION'
  | 'FINE_ALIGNMENT'
  | 'FSOC_ACTIVE'
  | 'RECOVERY_FAILED';

export type TurbulenceLevel = 'OFF' | 'LOW' | 'MEDIUM' | 'HIGH';
export type RfStatus = 'TRANSMITTING' | 'DISCOVERED' | 'IDLE';
export type TerminalRole = 'Target' | 'Other';
export type ScenarioId = 'normal' | 'temporary-loss' | 'deep-loss' | 'rogue-auth';

/** One remote terminal in the sector, as tracked by the simulator. */
export interface Terminal {
  id: string;
  name: string;
  /** "Detected" once the optical scan has swept over it. */
  status: string;
  rf_status: RfStatus;
  role: TerminalRole;
  position: Vec2;
  velocity: Vec2;
  heading: number;
  speed: number;
  trajectory: string;
  is_target: boolean;
}

/** Local ground station (Terminal A) pose and camera cone. */
export interface TerminalAState {
  x: number;
  y: number;
  orientation_deg: number;
  fov_deg: number;
}

/** Remote target terminal (Terminal B) pose and beacon. */
export interface TerminalBState {
  x: number;
  y: number;
  beacon_active: boolean;
  beacon_x: number;
  beacon_y: number;
}

/** World-space simulation snapshot used by the arena renderer. */
export interface ArenaState {
  terminal_a: TerminalAState;
  terminal_b: TerminalBState;
  terminals_list: Terminal[];
  world_width: number;
  world_height: number;
  kalman_pred: Vec2 | null;
  kalman_cov_trace: number;
  sim_step: number;
  scenario: ScenarioId;
  is_running: boolean;
  cloud_occlusion: boolean;
  vibration: number;
  turbulence: TurbulenceLevel;
}

/** HMAC-SHA256 challenge/response audit block (Part 3). */
export interface SecurityState {
  algorithm: string;
  challenge_nonce: string;
  hmac_sample: string;
  replay_protected: boolean;
  valid: boolean;
  stage: string;
  identification_status: string;
  target_terminal: string;
  target_terminal_name: string;
}

/** The full per-frame telemetry contract (`SystemTelemetry` on the Python side). */
export interface Telemetry {
  timestamp: number;
  current_state: SystemState;
  beacon_detected: boolean;
  beacon_position: Vec2 | null;
  tracking_status: string;
  prediction_status: string;
  optical_recovery_status: string;
  rf_status: string;
  terminal_id: string;
  authentication_status: string;
  approximate_direction: string | null;
  alignment_status: string;
  fsoc_status: string;
  error_status: string | null;
  camera_orientation_deg: number;
  camera_pan: number;
  camera_tilt: number;
  pointing_error_px: number | null;
  pointing_error_deg: number | null;
  tracking_error: number | null;
  rf_rssi: number | null;
  rf_direction: string | null;
  tracking_confidence: number;
  miss_count: number;
  recovery_attempt: number;
  terminal_world_position: Vec2 | null;
  terminal_velocity: Vec2 | null;
  terminal_heading: number;
  beacon_image_position: Vec2 | null;
  predicted_beacon_position: Vec2 | null;
  optical_signal_strength: number;
  optical_link_status: string;
  turbulence_level: TurbulenceLevel;
  vibration_level: number;
  cloud_occlusion: boolean;
  num_terminals: number;
  target_terminal: string;
  target_terminal_name: string;
  identification_status: string;
  security_algorithm: string;
  security_challenge_nonce: string;
  security_hmac_sample: string;
  security_replay_protected: boolean;
  security_valid: boolean;
  security_stage: string;
  recovery_mechanism: string;
  rf_discovery_state: string;
  rf_discovered_count: number;
  rf_total_count: number;
  rf_wave_active: boolean;
  optical_search_state: string;
  optical_tracking_state: string;
  fsoc_state: string;
  terminals_overview: Terminal[];
  recent_events: string[];
}

/** One SSE frame — the root object pushed on `/api/stream` at ~30 Hz. */
export interface TelemetryFrame {
  telemetry: Telemetry;
  arena: ArenaState;
  terminals: Terminal[];
  target_terminal: string;
  target_terminal_name: string;
  num_terminals: number;
  security: SecurityState;
  recovery_mechanism: string;
  events: string[];
  timestamp: number;
}

/** `/api/terminals` and `/api/terminals/config` response. */
export interface TerminalsConfig {
  num_terminals: number;
  target_terminal: string;
  available_pool: { id: string; name: string }[];
  overview: Terminal[];
}

/** `/api/disturbance` response — the simulator's authoritative echo. */
export interface DisturbanceState {
  cloud: boolean;
  cloud_occlusion: boolean;
  cloud_transit_frames: number;
  turbulence: TurbulenceLevel;
  turbulence_level: TurbulenceLevel;
  is_turbulent: boolean;
  vibration: number;
  vibration_amplitude: number;
  is_vibrating: boolean;
  severe: boolean;
}

export interface TestItem {
  name: string;
  status: 'PASSED' | 'FAILED' | 'PENDING' | 'RUNNING';
}

export interface TestCategory {
  id: string;
  title: string;
  passed: number;
  total: number;
  status: 'PASSED' | 'FAILED' | 'PENDING' | 'RUNNING';
  items: TestItem[];
}

/** `/api/tests/status` and `/api/tests/run` response. */
export interface TestStatus {
  is_running: boolean;
  current_step: string;
  progress_pct: number;
  total_tests: number;
  passed_tests: number;
  failed_tests: number;
  success: boolean;
  duration_sec: number;
  last_run_time: number;
  categories: TestCategory[];
}

/** `/api/action` response. */
export interface ActionResult {
  success: boolean;
  action: string;
  is_running: boolean;
  scenario: ScenarioId;
}
